README.md
# stepNCidx_client_JSON
Browser viewing of Step-NC files<BR><BR>
To run locally:<BR>
1. open nc-frames.html<BR>
2. in browser bar type "?foldername/"<BR> 
3. Make sure that the folder is in the same directory as the rest of the files<BR><BR><BR>


Full_JSON_Moldy has been given as a test folder to see the functionality.


NEWS
==
== ST-WebGL Viewer.
==

Changes in release 1.7 (2012-11-07)

* Added this NEWS file to release.

* UI tweaks to the part context menu

* Forced context menu to not move off the screen to the right.  
  (Bottom was already handled.)

* Changed red color for selected items to be a bit brighter

* Added save support for natsim server

Changes in release 1.6 (2012-08-14)

* Moved edit links functionality to the context menu.

* Added picking support.

* Moved all the STEP Tool deployment info (html and data) to stpidx.

* Updated NC code to use DataLoader

* Updated NC code to use collapsible HTML Tree.

* Renamed STEP-NC HTML to have nc- prefix.  (frames.html -> nc-frames.html,
 control.html -> nc-control.html)

* Sorted the items in the assembly tree. 

* Improved data loading code to be interruptable to avoid freezing the UI.

* Added visual feedback for selection

* Added event handling code to pop-down the menu when a click outside the menu
  is made

* Made visibility management code much more robust

* Highlighted element that menu was clicked over

* Moved the menu to place it on the screen, if necessary.

* Fixed bug where it was impossible to add to the set of loaded items when 
  starting from a save file.

* Expand the assembly tree as parts are loaded.

* Added "Loading" message when items are being loaded

* Added warning before a user loads a lot of data.

* Added links to STEP files in the data

* Fixed regression in STEP-NC data


Changes in release 1.5 (2012-07-03)

* Added support for incremetal loading.

* Added context menu to loading parts.

* Added selection feature to UI

Changes in release 1.4 (2012-05-14)

* Changed BG color for better contrast with dark shapes.

* Added code to collapse/expand assembly tree

* Made items in assembly tree gray if the geometry was missing.

Changes in release 1.4 (2012-04-25)

* Added "Save State" functionality to assembly viewer.

* Added Hypertext links to assembly nodes.

* Added hypertext link editor.

Changes in release 1.4 (2012-04-25)

* Removed annotations from bbox computation.

* Added assembly viewer

Changes in release 1.3 (2012-03-13)

* Fixed code to work where there is no setup (e.g. the moldy data set).

* Display polyline annotations.

Changes in release 1.2 (2012-03-07)

* Split modelview and project matrix in vertex shader.

* Updated shaders to use multiple, better lights to provide better contrast 
  for parallel faces.  In the process, moved most of the functionality to the
  fragment shader.

Changes in release 1.1 (2012-03-05)  (internal)

* Fixed regression in machine tool viewer where zoom +/- was not working

* Extended Shell object to support individual faces.  Earlier, we had assumed
  that there was a single <facets> element in the shell which contained all the
  facets for the shell.  We updated the file format to allow multiple 
  <facets> sets, each of which corresponds to a face, and each of which may have
  its own color.

* Added color support to shells

* Added Makefile targets to build zips and install stuff.

Changes in release 1.0 (2012-02-02)

* Move initialization to a timeout function to avoid a race conditions
  in Chrome where the window size comes up as 0x0, resulting in the
  canvas not getting propertly initialized.

* Added play button to automatically advance the tool along the toolpath.

* Calling it 1.0 since this represents the closure of the SBIR Pase 1 Option,
  and we now have a basic, functional tool.

Changes in release 0.6 (2012-02-01)

* Added controls to selectivly show/hide part, machine, tool, toolpath
  and fixture.

* Added UI to allow switching between several (or no) machine tools at
  runtime.

* Added machine drivers to allow machine tool visualization drived by
  STEP-NC data.

* Added support for fixtures.

* Refactored low-level code to use a unique ViewVolume for each machine 

* Update GeomView to make the ViewVolume a method on the scenegraph
  (optionally) for those cases where multiple ViewVolumes are
  required.  Simple cases (single instances) are still managed by the
  GeomView.

* Added unit conversion code to machine models and part

* Fixed shape bbox computation to use inverse transform.

Changes in release 0.5 (2012-01-03)

* Updated license/copyright header to BSD-style permissive license, as
  we use for demos

* Added support for displaying and controlling machines with
  rotational axes.

* Hide controls for axes that do not exist on a given machine.

* Updated behavior of +/- button on axis controls.  For rotational axes, moves
  axis by 5 degrees; for linear, moves it 2% of total axis distance.

Changes in release 0.5 (2011-12-21)

* Prettied up for web release

* Added machine tool viewer (machine-frames.html)

Changes in release 0.4 (2011-12-02)

* Removed STRLs from tree display.

Changes in release 0.3 (2011-12-01)

* Added code allow multiple to_be elements to be specified in the
  toolpath.

* Resolved the best to_be for the current WS.

* Internal changes to allow save and restore of color state

* Added suppoort for selective workplan elements.

* Moved color and transform state code into WebGLRenderingContext
  rather than their own methods.

Changes in release 0.2 (2011-11-22)

* Added code to display and move tool along toolpath

* Added user interface to update tool position.


Changes in release 0.1 (2011-11-18)

* Construct the workplan tree

* Display toolpaths on the impeller part.

* Created frames for workplan, control and 3d viewer.

* Added AP238 support


Changes in release 0.0 (2011-11-02)

* Got a rough demo that displays assemblies and allows interactive
  manipulation.


Shell.js
/* $RCSfile: Shell.js,v $
 * $Revision: 1.15 $ $Date: 2012/08/14 18:58:30 $
 * Auth: Jochen Fritz (jfritz@steptools.com)
 * 
 * Copyright (c) 1991-2012 by STEP Tools Inc. 
 * All Rights Reserved.
 * 
 * Permission to use, copy, modify, and distribute this software and
 * its documentation is hereby granted, provided that this copyright
 * notice and license appear on all copies of the software.
 * 
 * STEP TOOLS MAKES NO REPRESENTATIONS OR WARRANTIES ABOUT THE
 * SUITABILITY OF THE SOFTWARE, EITHER EXPRESS OR IMPLIED, INCLUDING
 * BUT NOT LIMITED TO THE IMPLIED WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE, OR NON-INFRINGEMENT. STEP TOOLS
 * SHALL NOT BE LIABLE FOR ANY DAMAGES SUFFERED BY LICENSEE AS A
 * RESULT OF USING, MODIFYING OR DISTRIBUTING THIS SOFTWARE OR ITS
 * DERIVATIVES.
 * 
 * 		----------------------------------------
 *
 * A 3D shell of triangles.  This will process the XML element
 * and build a Shell object for use by WebGL
 */

"use strict";

function Shell (builder, id) {

    this.use_count = 0;

    var ret = builder.make(id, this, "shell");
    if (ret)
	return ret;

    this.id = id;
    
    var el = builder.getElement(id);

    if (el == null)
	return null;

    var vtags = getObjects(el,'verts');
    if (vtags.length > 0) {
	this.load_all(el);
	return;
    }
    
    this.href = el.href;

    this.bbox
	= BoundingBox.fromArray(el["bbox"]);

    this.size = parseInt(el.size);
}


METHODS(Shell, {

    getName : function() {return "Shell";},

    loadData : function(tree, gl) {
	
	if (this.isLoaded())
	    return;

	this.gl = gl;
	
	var self = this;

	if (!this.href)
	    return;

	this.tree_el = tree;
	
	LOADER.addRequest(
	    this.href, this, 
	    function() {return self.getRank();} );
    },

    unloadData : function(gl) {
	this.loaded = false;
	this.loading = false;
	console.log("i made it here");
	if (!this.faces)
	    return;
	console.log("but not here");
	for (var i=0; i<this.faces.length; i++)
	    this.faces[i].unloadData(gl);

	delete this.color;
	delete this.faces;
    },

    getId : function() {return  this.id;},
    
    getCost : function() {
	return this.getFacetCount() * 3;
    },

    load_all : function(el) {
	var cont = this.loadElement(el);
	while (cont)
	    cont = cont();
    },

    load : function(json) {
	return this.loadElement(json);
    },
    
    loadElement : function(el) {
	
	var tree = this.tree_el;
	this.loading = true;
	
	if (tree) tree.expand(this);
	
	//console.log(el);
	var verts = getObjects(el['shell']['verts'], 'v');
	//console.log(verts);
	var points = new Float32Array(verts.length * 3);
	
	this.bbox = new BoundingBox();
	
	var color = el["color"];
	if (color)
	    this.color = parse_color(color);

	var faces = getObjects(el['shell'], 'facets');
	//console.log(faces);
	this.faces = [];
	//console.log(this.faces);
	return this.continue_load(verts, points, faces, 0);

    },

    continue_load : function(verts, points, faces, first) {
        //console.log("I made it here");
	if (!this.loading)
	    return;
	//console.log("but not here");
	var self = this;
	var i;


	var first_facet=0;
	if (first >= verts.length)
	    first_facet = first - verts.length
	else  {
	    var last_pt = first + 1000;
	    for (i=first; i<verts.length; i++) {
		if (i >= last_pt) {
		    return function() {
			return self.continue_load(verts, points, faces,
						  last_pt); 
		    }
		}
		
		var pt = verts[i]["p"];
		//console.log(pt);
		var coords = pt;
		//console.log(coords);
		
		if (coords.length != 3) {
		    throw new Error(
			"Must have exactly three coordinates in a point");
		}
		
		for (var j=0; j<3; j++) {
		    var v = parseFloat(coords[j]);
		    //console.log(v);
		    if (!isFinite(v))
			throw new Error("Number is not finite");
		    
		    this.bbox.updateI(j, v);
		    points[i * 3 + j] = v;
		    //console.log(points[i * 3 + j]);
		}
		//console.log(points);
	    }
	}
	    
	var facet_count = 0;

	for (i=first_facet; i<faces.length; i++) {
	    var face = new Face(this.gl, faces[i], points);
	    //console.log(face);
	    this.faces.push(face);

	    facet_count += face.facetCount();
	    if (facet_count > 1000) {
		return function() {
		    return self.continue_load(verts, points, faces,
					      verts.length+i+1); 
		}
	    }
	}

	this.loaded = true;
	this.loading = false;

	return null;
    },

    
    isLoaded : function() {
	return this.loaded;
    },
    
    getFacetCount : function() {
		
	if (this.size !== undefined)
	    return this.size;

	if (!this.faces) {
	    console.log ("No faces in shell: "+this.id);
	}

	console.log ("faces="+this.faces);
	
	this.size = 0;

	if (this.faces == null) return 0;
	
	for (var i=0; i<this.faces.length; i++) {
	    this.size += this.faces[i].facetCount();
	}

	console.log ("After loop");
	return this.size;
    },
    
    incrCount : function(loadables) {
	if (this.use_count == 0) {
	    if (!this.loaded)
		loadables.push(this);
	}
	this.use_count++
    },

    getRank : function() {
	if (this.rank) return this.rank;
	
	if (this.faces) {
	    this.rank = 0;
	    return 0;
	}

	if (!this.bbox)
	    return 0;
	
	var sz = this.size;
	if (sz == 0)
	    sz = 1.e20;

	var vol =
	    (this.bbox.maxx - this.bbox.minx) *
	    (this.bbox.maxy - this.bbox.miny) *
	    (this.bbox.maxz - this.bbox.minz);
	
	this.rank = this.use_count *vol / sz;

	return this.rank;
    },

    getLastDraw : function() {
	return this.last_draw;
    },
    
    draw : function(gl) {
	this.last_draw = gl.draw_serial;
	
	if (!this.faces)
	    return;
	
	var old_color = gl.saveColor();
	if ( this.color)
	    gl.setColorv(this.color);

	for (var i=0; i<this.faces.length; i++)
	    this.faces[i].draw(gl);
	
	gl.restoreColor(old_color);
    },

});


/*******************************/

function Face(gl, el, points)
{
    var color = el["color"];
    if (color)
	this.color = parse_color(color);
    
    var facets = getObjects(el, "f");

    var coords_buff = new Float32Array(facets.length * 9);
    var normals_buff = new Float32Array(facets.length * 9);

    this.facet_count = facets.length;
    
    for (var i=0; i<facets.length; i++) {
	var facet = facets[i];
	
	var idxs = facet["v"];
	//console.log(idxs);
	if (idxs.length != 3) 
	    throw new Error ("v elements does not have exactly 3 members");
	
	for (var j = 0; j < 3; j++) {
	    for (var k = 0; k < 3; k++) {
			coords_buff[i * 9 + j * 3 + k] = points[parseFloat(idxs[j]) * 3 + k];
	    }
	}

	var normals = facet['n'];

	if (normals.length != 3)
	    throw new Error("Expecting 3 n elements, have "+normals.length);
	    
	for (var j = 0; j < 3; j++) {
	    var coords = normals[j]["d"];
	    if (coords.length != 3)
		throw new Error ("Expecting 3 elements in the normal");
	    
	    for (var k = 0; k < 3; k++) {
	        var v = parseFloat(coords[k]);
		if (!isFinite(v))
		    throw new Error ("normal number is not finite");
		normals_buff[i*9 + j*3 + k] = v;
	    }
	}
    }

    if (gl) {
        this.init(gl, coords_buff, normals_buff);
    }
    else {
        this.coords = coords_buff;
        this.normals = normals_buff;
    }
    
}

METHODS(Face, {
    facetCount : function() {
	return this.facet_count;
    },

    unloadData : function(gl) {
	if (this.location_buff)
	    gl.deleteBuffer(this.location_buff);
	if (this.normal_buff)
	    gl.deleteBuffer(this.normal_buff);	
    },

    init : function(gl, coords, normals) {
	if (!this.location_buff) {
 	    this.location_buff = gl.createBuffer();
	    gl.bindBuffer(gl.ARRAY_BUFFER, this.location_buff);
	    gl.bufferData(gl.ARRAY_BUFFER, coords, gl.STATIC_DRAW);
	}

	if (!this.normal_buff) {
	    this.normal_buff = gl.createBuffer();
	    gl.bindBuffer(gl.ARRAY_BUFFER, this.normal_buff);
	    gl.bufferData(gl.ARRAY_BUFFER, normals, gl.STATIC_DRAW);	
	}		
    },
    
    draw : function(gl) {
	if (!this.location_buff || !this.normal_buff)
	    this.init(gl, this.coords, this.normals);
	
	gl.bindBuffer(gl.ARRAY_BUFFER, this.location_buff);
	gl.vertexAttribPointer(gl.pos_loc, 3, gl.FLOAT, false, 0, 0);
	
	gl.bindBuffer(gl.ARRAY_BUFFER, this.normal_buff);
	gl.vertexAttribPointer(gl.norm_loc, 3, gl.FLOAT, false, 0, 0);
	
	var count = this.facetCount() * 3;

	var old_color = gl.saveColor();
	if ( this.color)
	    gl.setColorv(this.color);
	
	gl.drawArrays(gl.TRIANGLES, 0, this.facetCount() *3 );

	gl.restoreColor(old_color);	
    },
});
